Fix Subset of a Subset not sliceable issue

TJ-coding opened the pull request:
Dataset can be indexed by a list, but a list can not be indexed by a list. This gives error when slicing a Subset initialised with a Subset, instead of a dataset.

Fixed the issue by changing the indices to a Tensor which can be indexed by a list.

Fixes #59512

TJ-coding committed:
Fix Subset of a Subset not sliceable issue

Dataset can be indexed by a list, but a list can not be indexed by a list. This gives error when slicing a Subset initialised with a Subset, instead of a dataset. Fixed by adding a feature that allow indexing a Subset by a list.

ejguan commented on torch/utils/data/dataset.py:
```diff
@@ -316,7 +316,10 @@ def __init__(self, dataset: Dataset[T_co], indices: Sequence[int]) -> None:
 
     def __getitem__(self, idx):
-        return self.dataset[self.indices[idx]]
+        if isinstance(self.indices, (list, tuple)) and isinstance(idx, Iterable):
+            return self.dataset[[self.indices[i] for i in idx]]
+        else:
+            return self.dataset[self.indices[idx]]
```
Let's just check `idx` is instance of list, because users may not want to iterate over tuple argument.  For other cases, we fall back to `self.dataset[self.indices[idx]]`

nit: you can remove `else` and return directly.

TJ-coding replied:
Sure makes sense, I will make the changes. Thank you for taking the time to review the code, I appreciate it.

ejguan reviewed (changes requested):
In general, it looks great.
Can you please add a test to `test/test_dataloader.py`?

TJ-coding commented:
> In general, it looks great.
> Can you please add a test to `test/test_dataloader.py`?

Sure, I will add a new test class in test_dataloader for Subset class.

TJ-coding committed:
Unittest for the new subset slicing feature

Added a unit test that checks if:
1. A Subset of a Dataset can be sliced given that the dataset is sliceable
2. A Subset of a Subset can be sliced given that the dataset of the root Subset is sliceable

ejguan commented on test/test_dataloader.py:
```diff
@@ -2274,6 +2274,47 @@ def test_conv_after_fork(self):
         for x in loader:
             self.assertEqual(x.shape, (1, 1, 1, 23999))
 
+class TestSubsetSlicing(TestCase):
```
Let's move them into https://github.com/pytorch/pytorch/blob/f9445c8a6b606dec9e59a8d8fd4e90a7d4e128d1/test/test_dataloader.py#L83

ejguan commented on test/test_dataloader.py:
```diff
@@ -2274,6 +2274,47 @@ def test_conv_after_fork(self):
+        # Testing slicing a subset initialized with a dataset
+        dataset = TensorDataset(torch.tensor([1, 2, 3, 4, 5]))
+        subset_of_dataset = Subset(dataset, [0, 1, 2, 3, 4])
+        try:
+            subset_of_dataset[:]
+        except Exception as e:
+            self.fail("A Subset object with a dataset attribute of a sliceable dataset must be sliceable but raised"
+                      + f"raised an exception: \n{e}")
```
We don't need to check this as we should expect it works.

ejguan commented on test/test_dataloader.py:
```diff
@@ -2274,6 +2274,47 @@ def test_conv_after_fork(self):
+        dataset = TensorDataset(torch.tensor([1, 2, 3, 4, 5]))
+        subset_of_dataset = Subset(dataset, [0, 1, 2, 3, 4])
+        subset_of_subset = Subset(subset_of_dataset, [0, 1, 2, 3, 4])
+        try:
+            subset_of_subset[:]
+        except Exception as e:
+            self.fail("A Subset object with a dataset attribute of another Subset object, with a dataset attribute of"
+                      + f"a sliceable Dataset, must be sliceable but raised an exception: \n{e}")
```
Ditto

ejguan commented on test/test_dataloader.py:
```diff
@@ -2274,6 +2274,47 @@ def test_conv_after_fork(self):
+        self.assertEqual(subset_of_subset[:], dataset[:])
+        self.assertEqual(subset_of_subset[0:-1:2], dataset[0:-1:2])
+
+        subset_of_subset.indices = [1, 3, 2]
+        self.assertEqual(subset_of_subset[:], dataset[[1, 3, 2]])
+
+        subset_of_subset.indices = torch.tensor([0, 1, 2, 3, 4])
+        self.assertEqual(subset_of_subset[:], dataset[:])
```
Ditto

ejguan commented on torch/utils/data/dataset.py:
```diff
@@ -316,6 +316,8 @@ def __init__(self, dataset: Dataset[T_co], indices: Sequence[int]) -> None:
         self.indices = indices
 
     def __getitem__(self, idx):
+        if isinstance(self.indices, list) and isinstance(idx, Iterable):
```
Let's change it to `if isinstance(idx, list):`. Let me know if you have any objection.

TJ-coding replied:
Simply removing the `isinstance(idx, Iterable)` will cause error, this is because `self.indices` may be a list but `idx` will not necessarily be an iterable. Passing a non iterable `idx` to the for loop in line 320 will cause error. E.g. slicing a subset  `subset[0:-1]` will give a sice object, `slice(0, -1, None)`, as `idx`  , which is not iterable. Indexing a subset would also give error since `subset[0]` will set ` idx` as 0 and this will also cause error in the line 320 for loop. But these non iterable `idx` can be handled by the original `return self.dataset[self.indices[idx]]`.

The feature line 319 and 320 add is the ability to index a subset by an Iterable E.g. `subset[0,1,2,3,4] `. This would have return an error in the previous implementation, and was causing the error when trying to slice a subset of a subset. E.g. if `idx` is a slice, `self.indices[idx]` is a tuple, an Iterable, hence line 321 would be `return self.dataset[**Iterable**]`. This would be ok if `self.dataset` is a Dataset since it can be indexed by an Iterable, but if `self.dataset` is a Subset then it will return error. Hence adding this feature solves the original problem of "Subset of a Subset not sliceable issue".

Sorry for the long comment, I hope it helps. Please let me know if there are any problems or changes that should be made. Thank you.

ejguan replied:
I am referring the condition that `idx` is a `list` (not `indices` is `list`). We should expect `indices` always being `list`.

ejguan replied:
In general, I mean we can delete the check for `isinstance(self.indices, list)`, which should always be true.

TJ-coding replied:
I see, I am sorry for the miss understanding. The reason why I left the list there is because I thought someone may want to use a numpy array or integer pytorch tensor as an indices instead, which already supports being indexed by an Iterable, and they may be faster than doing a python for loop. But sure, I don't have any other reasonable objection against removing it so I am very much happy to remove it.

ejguan replied:
IMO, the reason is kind simple. Subset is normally used by `random_split`. If users want to use different ways to create Subset, it can be easily achieved to write a custom class. It's not possible for us to support all the user behavior for a single class.
BTW, please also switch to check `idx` against `list` rather than `Iterable`. If `idx` is a tuple, it should not fall into this logic.

TJ-coding replied:
Sure, I completely agree with you. I had a wrong idea on how indexing by tuple behaves on a torch tensors. I think the interface of Subset should be similar to of Dataset with a polymorphism like concept.

ejguan reviewed (changes requested):
Few minor comments. Can you please rebase this PR to the branch of viable/strict?

ejguan reviewed (approved):
LGTM. Thanks for submitting the PR.
